File: src/dp/lc337.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function rob(root: TreeNode | null): number {
  const trueMap = new Map<TreeNode, number>();
  const falseMap = new Map<TreeNode, number>();
  const dfs = (cur: TreeNode | null, useParent: boolean): number => {
    if (cur === null) {
      return 0;
    }
    if (useParent) {
      if (trueMap.has(cur)) {
        return trueMap.get(cur) ?? NaN;
      }
      const res = dfs(cur.left, false) + dfs(cur.right, false);
      trueMap.set(cur, res);
      return res;
    } else {
      if (falseMap.has(cur)) {
        return falseMap.get(cur) ?? NaN;
      }
      const res = Math.max(
        cur.val + dfs(cur.left, true) + dfs(cur.right, true),
        dfs(cur.left, false) + dfs(cur.right, false),
      );
      falseMap.set(cur, res);
      return res;
    }
  };
  return dfs(root, false);
}

export default rob;

File: src/dp/lc337.cc
#include <algorithm>
#include <functional>
#include <unordered_map>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : val(x), left(left), right(right) {}
};

using namespace std;

class Solution {
 public:
  int rob(TreeNode* root) {
    auto parentIdxs = vector<int>();
    auto trueMap = unordered_map<TreeNode*, int>();
    auto falseMap = unordered_map<TreeNode*, int>();
    function<int(TreeNode * cur, bool useParent)> dfs;
    dfs = [&](TreeNode* cur, bool useParent) -> int {
      if (cur == nullptr) {
        return 0;
      }
      if (useParent) {
        if (trueMap.find(cur) != trueMap.end()) {
          return trueMap[cur];
        }
        auto res = dfs(cur->left, false) + dfs(cur->right, false);
        trueMap[cur] = res;
        return res;
      } else {
        if (falseMap.find(cur) != falseMap.end()) {
          return falseMap[cur];
        }
        auto res = max(cur->val + dfs(cur->left, true) + dfs(cur->right, true),
                       dfs(cur->left, false) + dfs(cur->right, false));
        falseMap[cur] = res;
        return res;
      };
    };
    return dfs(root, false);
  }
};
